lib/SILOptimizer/IPO/CrossModuleOptimization.cpp

```diff
@@ -266,31 +264,81 @@ void CrossModuleOptimization::serializeFunctionsInModule(
+                      [&](auto entry) {
+        return entry.getImplementation();
+      });
+      trySerializeFunctions(classMethodsToSerialize);
+
+      bool containsInternal = llvm::any_of(classMethodsToSerialize, [&](auto &method) {
+        // If the entry is internal, vtable should not be serialized.
+        // However, if the entry is not serialized but has the right
+        // visibility, it can still be referenced, thus the vtable
+        // should serialized.
+        return !method->hasValidLinkageForFragileRef(getRightSerializedKind(M));
+      });
+      if (!containsInternal)
+        vt->setSerializedKind(getRightSerializedKind(M));
     }
   }
 
+  // Witness thunks are not serialized, so serialize them here.
   for (auto &wt : M.getWitnessTables()) {
-    if (wt.isNotSerialized() && 
+    if (wt.getSerializedKind() != getRightSerializedKind(M) &&
         hasPublicOrPackageVisibility(wt.getLinkage(), /*includePackage*/ true)) {
-      for (auto &entry : wt.getEntries()) {
-        // Witness thunks are not serialized, so serialize them here.
-        if (entry.getKind() == SILWitnessTable::Method &&
-            entry.getMethodWitness().Witness->isNotSerialized() &&
-            isSerializeCandidate(entry.getMethodWitness().Witness,
-                                 M.getOptions())) {
-          entry.getMethodWitness().Witness->setSerializedKind(getRightSerializedKind(M));
-        }
-      }
-      // Then serialize the witness table itself.
-      wt.setSerializedKind(getRightSerializedKind(M));
+      // This checks if a wtable entry is not serialized and attempts to
+      // serialize (and its references) if they have the right visibility.
+      // This should not be necessary but is added to ensure all applicable
+      // symbols are serialized. Whether serialized or not is cached so
+      // this check shouldn't be expensive.
+      auto unserializedWTMethodRange = llvm::make_filter_range(wt.getEntries(), [&](auto &entry) {
+        return entry.getKind() == SILWitnessTable::Method &&
+        entry.getMethodWitness().Witness->getSerializedKind() != getRightSerializedKind(M);
+      });
+      std::vector<SILFunction *> wtMethodsToSerialize;
+      llvm::transform(unserializedWTMethodRange,
+                      std::back_inserter(wtMethodsToSerialize),
+                      [&](auto entry) {
+        return entry.getMethodWitness().Witness;
+      });
+      trySerializeFunctions(wtMethodsToSerialize);
+
+      bool containsInternal = llvm::any_of(wtMethodsToSerialize, [&](auto &method) {
+        // If the entry is internal, wtable should not be serialized.
+        // However, if the entry is not serialized but has the right
+        // visibility, it can still be referenced, thus the vtable
+        // should serialized.
+        return !method->hasValidLinkageForFragileRef(getRightSerializedKind(M));
+      });
+      if (!containsInternal)
+        wt.setSerializedKind(getRightSerializedKind(M));
```

Comment:
the logics for serializing v-table and w-table are very much alike. Can we extract a common function to handle both?

Reply:
Some of these checks might not be necessary but added as a caution as mentioned in the comments; will follow up and remove them if found to be unnecessary.